graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -451,127 +455,215 @@ def remove_backtrack_blocks_from_fold(folded_ir_blocks):
-        new_traverse = []
-        for step in traverse:
-            if not isinstance(step.root_block, Traverse) or not step.root_block.optional:
-                new_traverse.append(step)
-            else:
-                field_name = step.root_block.get_field_name()
-                new_predicate = filter_local_field_existence(field_name)
-                old_filter = new_traverse[-1].where_block
-                if old_filter:
-                    new_predicate = BinaryComposition(u'&&', old_filter.predicate, new_predicate)
-                new_traverse[-1] = new_traverse[-1]._replace(where_block=Filter(new_predicate))
-                return new_traverse
-        raise AssertionError(u"No optional traverse found in: {}".format(traverse))
-
-    def make_mandatory_traverse(traverse):
-        """Return a new traversal, without the optional condition."""
-        new_traverse = []
-        for step in traverse:
-            if not isinstance(step.root_block, Traverse) or not step.root_block.optional:
-                new_traverse.append(step)
-            else:
-                old_block = step.root_block
-                new_root_block = Traverse(old_block.direction, old_block.edge_name)
-                new_traverse.append(step._replace(root_block=new_root_block))
-        return new_traverse
-
-    match_traversals_possibilities = []
-    for traversal in match_query.match_traversals:
-        if is_optional_traverse(traversal):
-            current_posibilities = [
-                remove_optional_traverse(traversal),
-                make_mandatory_traverse(traversal)
-            ]
-        else:
-            current_posibilities = [traversal]
-        match_traversals_possibilities.append(current_posibilities)
-    compound_match_traversals = list(itertools.product(*match_traversals_possibilities))
+    return CompoundMatchQuery(match_queries=match_queries)
 
-    return CompoundMatchQuery(
-        match_queries=[
-            MatchQuery(
-                match_traversals=match_traversals,
-                folds=match_query.folds,
-                output_block=match_query.output_block,
-            )
-            for match_traversals in compound_match_traversals
-        ]
-    )
+
+def _get_present_locations_from_match_traversals(match_traversals):
+    """Return the set of locations and non-optional locations present in the given match traversals.
+
+    When enumerating the possibilities for optional traversals,
+    the resulting match traversals may have sections of the query omitted.
+    These locations will not be included in the returned `present_locations`.
+    All of the above locations that are not optional traverse locations
+    will be included in present_non_optional_locations.
+
+    Args:
+        match_traversals: one possible list of match traversals generated from a query
+            containing @optional traversal(s)
+
+    Returns:
```

Comment:
You need to state that the return type is a tuple `(present_locations, present_non_optional_locations)`.